# Computer Assignment 1

Negar Moradi 810198543
    

## Introduction

#### The purpose of this project is to solve the given problem using search algorithms and determine the differences between uninformed search algorithms (BFS and IDS) and informed search algorithms (A* and weighted A*).

#### At first, we have a table n * m. We may have a potion or a doubling drug in each box. Starting at 0,0, we should go to n-1,m-1.

#### initial state: Each potion is in its own box and the doctor is in box (0,0).                                                                                              The goal is to drink all the potions with minimum cost. 

##### First we define a function to get input. This function builds our table and holds the coordinates of each potion, drug and block. We also have functions to convert a string to a list and a list to a string (this function is used to convert states to sets).
##### We use the success function to determine whether we have reached our goal. 


In [1]:
from time import time
from collections import deque
from math import sqrt
from heapq import heappush
from heapq import heappop


filenames = ['test1.in', 'test2.in', 'test3.in']

def get_input(testname):

    global m
    global n

    f = open(testname, "r")
    nandm = list(f.readline().split(' '))
    n, m = (int(nandm[0]),int(nandm[1]))
    candk = list(f.readline().split(' '))
    c, k = (int(candk[0]),int(candk[1]))

    potions = []
    Ddrug = []
    blocked = []

    for i in range(c):
        potions.append(list(map(int,f.readline().split(' '))))
    
    for i in range(k):
        Ddrug.append(list(map(int,f.readline().split(' '))))

    d = int(f.readline())

    for i in range(d):
        blocked.append(list(map(int, f.readline().split(' '))))

    return potions, Ddrug, blocked

def success(di:list(), pot:list()):
    flag = False
    for i in di:
        if(i[0]!=n-1 or i[1]!=m-1):
            flag = True
            break
    return (flag == False and len(pot) == 0)


def changetos(l):
    s = ''
    if(len(l)==0):
        return s
    for i in l:
        s+=str(i[0])+','+str(i[1])+'-'
    return s[:-1]

def changetol(s:str):
    l = []
    if(len(s)==0):
        return l
    ss = s.split('-')
    r = []
    for i in ss:
        r.append(list(map(int,i.split(','))))
    return r




## BFS

##### Doctors, potions, and medicines are stored in a set().  For each doctor, we check all 4 directions (up and down, left and right) and look for potions using BFS. Then we update our variables based on whether the potion or medicine is in each box.
##### To implement the frontier, we use a queue and to explore, we use a set.
##### The par dictionary contains the parent of each state. The all_states variable stores all the states we met
#####  When cost is 1, BFS is optimal.                                                                                                                                                                                     time complexity: O(b^d)                                                                                                                                                                                                           space complexity: O(b^d)  

In [2]:

all_state = 0

def bfs(doc, potions, medicines, blocked:list):
    
    global all_state

    starts = (doc, changetos(potions),changetos(medicines))

    frontier = deque()
    frontier.appendleft(starts)
    explored = set()
    par=dict()

    if(success(doc, potions)):
        return par, starts, starts
    

    while(True):

        if(not frontier):
            return False

        cur = frontier.pop()
        if(cur in explored):
            continue
        explored.add(cur)

        doc,pot,med = cur
        doc,pot,med = changetol(doc),changetol(pot),changetol(med)

        for i in (doc):

            x ,y = i[0],i[1]
        #in pot
            if ([x,y] in pot):
                pot.remove([x,y])
        #in med
            if ([x,y] in med):
                doc.append([n-1,0])
                med.remove([x,y])
        #goal
            if(x == n-1 and y == m-1):
                if(success(doc, pot)):
                    return par, cur, starts
                doc.remove([x,y])
                continue

            #down
            if (x - 1 >= 0):
                if([x-1,y] not in blocked):
                    i[0] = x - 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[0] = x
                    if(not new_state in explored):
                        par[new_state]=cur
                        if(success(changetol(ndoc), pot)):
                            return par, new_state, starts
                        frontier.appendleft(new_state)
                        all_state += 1
                    
                   


            #up
            if (x + 1 < n):
                if([x+1,y] not in blocked):
                    i[0] = x + 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[0] = x
                    if(not new_state in explored):
                        par[new_state]=cur
                        if(success(changetol(ndoc), pot)):
                            return par, new_state, starts
                        frontier.appendleft(new_state)
                        all_state += 1
                
                

            #left
            if (y - 1 >= 0):
                if([x,y-1] not in blocked):
                    i[1] = y - 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[1] = y
                    if(not new_state in explored):
                        par[new_state]=cur
                        if(success(changetol(ndoc), pot)):
                            return par, new_state, starts
                        frontier.appendleft(new_state)
                        all_state += 1
                

            #right
            if (y + 1 < m):
                if([x,y+1] not in blocked):
                    i[1] = y + 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[1] = y
                    if( not new_state in explored):
                        par[new_state]=cur
                        if(success(changetol(ndoc), pot)):
                            return par, new_state, starts
                        frontier.appendleft(new_state)
                        all_state += 1
                


In [3]:
def print_bfs(x,snode):
    
    if(x==snode):
        return
    doc, pot, med = x
    pathp.append(doc)
    print_bfs(path[x],snode)

def printpath_bfs(filename):

    global path
    global pathp

    all_state = 0
    pathp = []
    t1 = time()
    potions, Ddrug, blocked = get_input(filename)
    doctors = '0,0'
    path, node, starts = bfs(doctors,potions, Ddrug, blocked)
    print_bfs(node,starts)
    pathp.reverse()
    print(*pathp, sep='\n')
    t2 = time()
    print(t2-t1)
    return len(pathp)




print("result for BFS:")

for i in range(len(filenames)):
    print('test' + str(i+1))
    dist = printpath_bfs(filenames[i])
    print('Number of visited states:' + str(all_state))
    print('Distance:' + str(dist))
    print("-------------------------------------------------------------------------------------------------------------------")
        


result for BFS:
test1
0,1
0,1-2,0
0,1-2,1
0,1-2,2
0,1-2,3
0,1-3,3
0,2-3,3
0,3-3,3
1,3-3,3
2,3-3,3
3,3-3,3
0.08395171165466309
Number of visited states:2305
Distance:11
-------------------------------------------------------------------------------------------------------------------
test2
0,0-2,0
0,0-2,1
0,0-2,2
0,0-2,3
0,0-3,3
0,1-3,3
0,2-3,3
1,2-3,3
2,2-3,3
3,2-3,3
3,3-3,3
13.402283668518066
Number of visited states:529701
Distance:11
-------------------------------------------------------------------------------------------------------------------
test3
1,0
2,0
2,1
2,2
2,2-5,1
2,2-5,2
2,2-5,3
2,2-4,3
2,2-4,4
2,2-4,5
2,2-5,5
1,2-5,5
1,3-5,5
1,4-5,5
1,5-5,5
2,5-5,5
3,5-5,5
4,5-5,5
5,5-5,5
9.739392042160034
Number of visited states:915288
Distance:19
-------------------------------------------------------------------------------------------------------------------


### Advantages:

This algorithm is complete, so if a solution can be found it will be found. <br>
The solution is optimal.

### Disadvantages:

It takes a lot of memory to run this algorithm. <br>
When the optimal answer has a big depth, it will take a long time to find

## IDS

###### In a loop, we give a depth to the DFS function and based on that depth, we move forward until we find the solution. The DFS function takes four arguments: node, nodeparent, depth, and limit and returns one node(final node). A node is a set of states. In this part we use the explored function to store depths.
###### For each doctor, we check all 4 directions (up and down, left and right) and look for potions using DFS. Then we update our variables based on whether the potion or medicine is in each box.  Each series only goes to a depth equal to the size of the limit. If we reach the goal, we return node.
##### To implement the explored, we use a set. The all_states variable stores all the states we met
##### The pathnode dictionary contains the parent of each state.
###### We would return false if we had searched all the depths and reach the limit but still didn't find a solution.
##### Besides not having new_state in explored, we have another condition. We need to check DFS again if the depth of the current state is less than the depth of the new state because we came to this state from a shorter path.



In [4]:
all_states = 0
pathnode = dict() 
explored = dict()
mark = dict()


def dfs(stnode:set(), parent:set(), dep, limit):

    if(dep == limit):
       return False

    global all_states

    all_states = 1 + all_states

    pathnode[stnode] = parent

    if(stnode != parent):
        explored[stnode] = explored[parent] + 1

    
    doc,pot,med = stnode
    doc,pot,med= changetol(doc),changetol(pot),changetol(med)

    mark[stnode] = 1

    if(success(doc, pot)):
        return stnode
    

        

    for i in (doc):

        x ,y = i[0],i[1]
    #in pot
        if ([x,y] in pot):
            pot.remove([x,y])
    #in med
        if ([x,y] in med):
            doc.append([n-1,0])
            med.remove([x,y])
            pathnode[(changetos(doc),changetos(pot),changetos(med))]=parent
    #goal
        if(x == n-1 and y == m-1):
            if(success(doc, pot)):
                return stnode
            doc.remove([x,y])
            continue

        #down
        if (x - 1 >= 0):
            if([x-1,y] not in blocked):
                i[0] = x - 1
                npot = changetos(pot)
                nmed = changetos(med)
                ndoc = changetos(doc)
                new_state = (ndoc, npot, nmed)
                i[0] = x
                if(not new_state in mark or dep+1<explored[new_state]):
                    check = dfs(new_state, stnode, dep+1, limit)
                    if(check):
                        if(not new_state in mark):
                            del mark[stnode]
                        return check
                        


            #up
        if (x + 1 < n):
            if([x+1,y] not in blocked):
                i[0] = x + 1
                npot = changetos(pot)
                nmed = changetos(med)
                ndoc = changetos(doc)
                new_state = (ndoc, npot, nmed)
                i[0] = x
                if(not new_state in mark or dep+1<explored[new_state]):
                    check = dfs(new_state, stnode, dep+1, limit)
                    if(check):
                        if(not new_state in mark):
                            del mark[stnode]
                        return check
                        


            #left
        if (y - 1 >= 0):
            if([x,y-1] not in blocked):
                i[1] = y - 1
                npot = changetos(pot)
                nmed = changetos(med)
                ndoc = changetos(doc)
                new_state = (ndoc, npot, nmed)
                i[1] = y
                if(not new_state in mark or dep+1<explored[new_state]):
                    check = dfs(new_state, stnode, dep+1, limit)
                    if(check):
                        if(not new_state in mark):
                            del mark[stnode]
                        return check
                        


            #right
        if (y + 1 < m):
            if([x,y+1] not in blocked):
                i[1] = y + 1
                npot = changetos(pot)
                nmed = changetos(med)
                ndoc = changetos(doc)
                new_state = (ndoc, npot, nmed)
                i[1] = y
                if(not new_state in mark or dep+1<explored[new_state]):
                    check = dfs(new_state, stnode, dep+1, limit)
                    if(check):
                        if(not new_state in mark):
                            del mark[stnode]
                        return check
        
    return False
                           

def IDS(doc, potions, medicines, blocked:list):
    maxx = 1000
    starts = (doc, changetos(potions),changetos(medicines))
    explored[starts]=0
    for dss in range(100):
        mark.clear()
        pathnode.clear()
        node = dfs(starts,starts,0, dss)
        if(node):
            return node,starts , dss
    return False, False



In [5]:

def print_IDS(x,snode):
    
    if(x==snode):
        return
    doc, pot, med= x
    pathp.append(doc)
    print_IDS(pathnode[x],snode)


def printpath_IDS(filename):

    global path
    global pathp
    global blocked

    pathp = []
    t1 = time()
    potions, Ddrug, blocked = get_input(filename)
    doctors = '0,0'
    node, starts, distance = IDS(doctors,potions, Ddrug, blocked)
    if(node!= False):
        print_IDS(node,starts)
    pathp.reverse()
    print(*pathp, sep='\n')
    t2 = time()
    print(t2-t1)
    print('Number of visited states:' + str(all_states))
    print('Distance:' + str(distance-1))
    
print("result for IDS:")

for i in range(len(filenames)):
    print('test' + str(i+1))
    printpath_IDS(filenames[i])
    print("-------------------------------------------------------------------------------------------------------------------")
   

result for IDS:
test1
0,1
0,2-3,0
0,3-3,0
1,3-3,0
2,3-3,0
2,3-2,0
2,3-3,0
2,3-3,1
2,3-3,2
2,3-3,3
3,3-3,3
0.3437988758087158
Number of visited states:7289
Distance:11
-------------------------------------------------------------------------------------------------------------------
test2
0,1-3,0
0,2-3,0
1,2-3,0
2,2-3,0
3,2-3,0
3,2-2,0
3,2-2,1
3,2-2,2
3,2-3,2
3,2-3,3
3,3-3,3
38.84763526916504
Number of visited states:378611
Distance:11
-------------------------------------------------------------------------------------------------------------------
test3
1,0
2,0
2,1
2,2
1,2-5,0
1,3-5,0
1,4-5,0
1,5-5,0
2,5-5,0
3,5-5,0
4,5-5,0
4,5-5,1
4,5-5,2
4,5-5,3
4,5-4,3
4,5-5,3
4,5-5,4
4,5-5,5
5,5-5,5
50.10417151451111
Number of visited states:954998
Distance:19
-------------------------------------------------------------------------------------------------------------------


### Advantages:

This algorithm is also complete, so if a solution can be found, it will be found. <br>
When step cost is 1,this algorithm is also optimal.     

### Disadvantages:

The algorithm is slow, and it counts many duplicate states.

## A* 

###### First we define the heuristic() function. The function finds the closest doctor to each potion (d), and returns the potion distance that has the most d. (It finds the potion with the greatest distance from the other doctors.)
##### For a consistent heuristic, the solution is optimal. Our heuristic function is consistent since it is always non-increasing in its path. 

In [6]:
#in A* alpha is 1

def heuristic(doc, pot, alpha):

    minpot = [0]
    docd = []
    for j in pot:
        for i in doc:
            x = sqrt((j[0]-i[0])**2 + (j[1]-i[1])**2)
            docd.append(x)
        minpot.append(min(docd))

    return max(minpot)*alpha

###### This function is implemented just like BFS, but the difference is that we always pick the state with the lowest heuristic() value when we pop from the frontier list. In our state variable, we also store the value of heuristic for each node. 
###### For each doctor, we check all 4 directions (up and down, left and right) and look for potions using BFS. Then we update our variables based on whether the potion or medicine is in each box.
###### To implement the frontier, we use a list(heap) and to explore, we use a set.
##### The par dictionary contains the parent of each state. The all_states variable stores all the states we met

In [7]:
           
frontier = []
explored = set()
all_states = int()
par=dict()
mark = dict()

def A_Star(doc, potions, medicines, blocked:list, alpha):
    
    starts = (doc, changetos(potions),changetos(medicines))

    global all_states

    if(success(doc, potions)):
        return starts, starts
    
    heurs = heuristic([[0,0]],potions,alpha)
    #g(n)+h(n)
    heappush(frontier, (0+heurs,starts))
    mark[(0+heurs,starts)] = 1


    while(True):

        if(not frontier):
            return 

        cur1 = heappop(frontier)

        cur, curs = cur1[0],cur1[1]

        if(curs in explored):
            continue

        del mark[cur1]

        doc,pot,med = curs
        doc,pot,med = changetol(doc),changetol(pot),changetol(med)

        if success(doc, pot):
            return curs, starts

        explored.add(curs)

        for i in (doc):

            x ,y = i[0],i[1]
        #in pot
            if ([x,y] in pot):
                pot.remove([x,y])
        #in med
            if ([x,y] in med):
                doc.append([n-1,0])
                med.remove([x,y])
        #goal
            if(x == n-1 and y == m-1):
                if(success(doc, pot)):
                    return  curs, starts
                doc.remove([x,y])
                continue

        #down
            if (x - 1 >= 0):
                if([x-1,y] not in blocked):
                    i[0] = x - 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[0] = x

                    if(not new_state in explored):
                        new_hg = cur + 1 + heuristic(changetol(ndoc),pot, alpha) 
                        if(not new_hg in mark):
                            par[new_state]=curs
                            if(success(changetol(ndoc), pot)):
                                return  new_state, starts
                            heappush(frontier, (new_hg,new_state))
                            mark[(new_hg,new_state)] = 1
                            all_states += 1


        #up
            if (x + 1 < n):
                if([x+1,y] not in blocked):
                    i[0] = x + 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[0] = x
                    if(not new_state in explored):
                        new_hg = cur + 1 + heuristic(changetol(ndoc),pot, alpha)
                        if(not new_hg in mark):
                            par[new_state]=curs
                            if(success(changetol(ndoc), pot)):
                                return  new_state, starts
                            heappush(frontier, (new_hg,new_state))
                            mark[(new_hg,new_state)] = 1
                            all_states += 1

        #left
            if (y - 1 >= 0):
                if([x,y-1] not in blocked):
                    i[1] = y - 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[1] = y
                    if(not new_state in explored):
                        new_hg = cur + 1 + heuristic(changetol(ndoc),pot, alpha) 
                        if(not new_hg in mark):
                            par[new_state]=curs
                            if(success(changetol(ndoc), pot)):
                                return  new_state, starts
                            heappush(frontier, (new_hg,new_state))
                            mark[(new_hg,new_state)] = 1
                            all_states += 1

        #right
            if (y + 1 < m):
                if([x,y+1] not in blocked):
                    i[1] = y + 1
                    npot = changetos(pot)
                    nmed = changetos(med)
                    ndoc = changetos(doc)
                    new_state = (ndoc, npot, nmed)
                    i[1] = y
                    if(not new_state in explored):
                        new_hg = cur + 1 + heuristic(changetol(ndoc),pot, alpha) 
                        if(not new_hg in mark):
                            par[new_state]=curs
                            if(success(changetol(ndoc), pot)):
                                return  new_state, starts
                            heappush(frontier, (new_hg,new_state))
                            mark[(new_hg,new_state)] = 1
                            all_states += 1

###### For A* we use alpha = 1

In [8]:

def print_A_star(x,snode):
    
    if(x==snode):
        return
    doc, pot, med = x
    pathp.append(doc)
    print_A_star(par[x],snode)


      
def printpath_A_star(filename, alpha):

    global path
    global pathp

    pathp = []
    t1 = time()
    potions, Ddrug, blocked = get_input(filename)
    doctors = '0,0'
    node, starts = A_Star(doctors,potions, Ddrug, blocked, alpha)
    print_A_star(node,starts)
    pathp.reverse()
    print(*pathp, sep='\n')
    t2 = time()
    print('Number of visited states: ' + str(all_states))
    print('Distance: ' + str(len(pathp)))
    print('Time: ' + str(t2-t1))


print("result for A*:")

for i in range(len(filenames)):
    print('test' + str(i+1))
    printpath_A_star(filenames[i],1)
    mark.clear()
    par.clear()
    explored.clear()
    frontier.clear()
    all_states = 0
    print("-------------------------------------------------------------------------------------------------------------------")
        


result for A*:
test1
0,1
0,1-2,0
0,2-2,0
0,3-2,0
1,3-2,0
2,3-2,0
2,3-3,0
2,3-3,1
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 970
Distance: 11
Time: 0.025983333587646484
-------------------------------------------------------------------------------------------------------------------
test2
0,0-2,0
0,1-2,0
0,2-2,0
1,2-2,0
2,2-2,0
2,3-2,0
2,3-2,1
2,3-2,2
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 66528
Distance: 11
Time: 2.2946951389312744
-------------------------------------------------------------------------------------------------------------------
test3
1,0
2,0
2,1
2,2
1,2-5,0
1,3-5,0
1,4-5,0
1,5-5,0
1,5-5,1
1,5-5,2
1,5-5,3
1,5-4,3
2,5-4,3
3,5-4,3
4,5-4,3
4,5-5,3
4,5-5,4
4,5-5,5
5,5-5,5
Number of visited states: 58838
Distance: 19
Time: 2.4515891075134277
-------------------------------------------------------------------------------------------------------------------


### Advantages:

If the heuristic function is consistent, this algorithm is optimal. <br>
less memory and faster

## Weighted A*

##### For this part, we use the same functions as above, but in the heuristic function, we multiply the final answer by alpha.

### For $\alpha = 1.5$

In [9]:

print("result for weghted A* with alpha=1.5:")

for i in range(len(filenames)):
    print('test' + str(i+1))
    printpath_A_star(filenames[i],1.5)
    mark.clear()
    par.clear()
    explored.clear()
    frontier.clear()
    all_states = 0
    print("-------------------------------------------------------------------------------------------------------------------")
        

result for weghted A* with alpha=1.5:
test1
0,1
0,2-3,0
0,2-2,0
0,3-2,0
1,3-2,0
2,3-2,0
2,3-3,0
2,3-3,1
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 801
Distance: 11
Time: 0.03697967529296875
-------------------------------------------------------------------------------------------------------------------
test2
0,1-3,0
0,2-3,0
1,2-3,0
1,2-2,0
2,2-2,0
2,3-2,0
2,3-2,1
2,3-2,2
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 44560
Distance: 11
Time: 1.448164701461792
-------------------------------------------------------------------------------------------------------------------
test3
1,0
2,0
2,1
2,2
1,2-5,0
1,3-5,0
1,4-5,0
1,5-5,0
2,5-5,0
3,5-5,0
3,5-5,1
3,5-5,2
3,5-5,3
3,5-4,3
4,5-4,3
4,5-5,3
4,5-5,4
4,5-5,5
5,5-5,5
Number of visited states: 38679
Distance: 19
Time: 1.620067834854126
-------------------------------------------------------------------------------------------------------------------


### For $\alpha = 3$

In [10]:

print("result for weighted A* with alpha=3:")

for i in range(len(filenames)):
    print('test' + str(i+1))
    printpath_A_star(filenames[i],3)
    mark.clear()
    par.clear()
    explored.clear()
    frontier.clear()
    all_states = 0
    print("-------------------------------------------------------------------------------------------------------------------")
        

result for weighted A* with alpha=3:
test1
0,1
0,2-3,0
0,2-2,0
0,3-2,0
1,3-2,0
2,3-2,0
2,3-3,0
2,3-3,1
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 665
Distance: 11
Time: 0.03398013114929199
-------------------------------------------------------------------------------------------------------------------
test2
0,1-3,0
0,2-3,0
1,2-3,0
1,2-2,0
2,2-2,0
2,3-2,0
2,3-2,1
2,3-2,2
2,3-3,2
2,3-3,3
3,3-3,3
Number of visited states: 37167
Distance: 11
Time: 1.2153003215789795
-------------------------------------------------------------------------------------------------------------------
test3
1,0
2,0
2,1
2,2
1,2-5,0
1,3-5,0
1,4-5,0
1,5-5,0
2,5-5,0
3,5-5,0
4,5-5,0
4,4-5,0
4,3-5,0
4,3-5,1
4,4-5,1
4,4-5,2
4,4-5,3
4,4-5,4
4,4-5,5
4,5-5,5
5,5-5,5
Number of visited states: 31118
Distance: 21
Time: 1.0933783054351807
-------------------------------------------------------------------------------------------------------------------


### Advantages:

Less memory <br>
Faster 

### Disadvantages:

The result may not be optimal any more.

#### Test1:
    
|       |cost     |Number of visited states|average time|
|:-----:|:-------:|:----------------------:|:----------:|
|BFS    |11       |2305                    |0.049       |
|IDS    |11       |7289                    |0.265       |
|A*     |11       |970                     |0.029        |
|weighted A*,$\alpha$ = 1.5|11|801         |0.027        |
|weighted A*,$\alpha$ = 3|11|665           |0.022       |

#### Test2:
    
|       |cost     |Number of visited states|average time|
|:-----:|:-------:|:----------------------:|:----------:|
|BFS    |11       |527396                  |10.50       |
|IDS    |11       |371322                  |36.65       |
|A*     |11       |66528                   |2.02        |
|weighted A*,$\alpha$ = 1.5|11|44560       |1.33        |
|weighted A*,$\alpha$ = 3|11|37167         |1.08        |

#### Test3:
    
|       |cost     |Number of visited states|average time|
|:-----:|:-------:|:----------------------:|:----------:|
|BFS    |19       |385587                  |8.12       |
|IDS    |19       |576377                  |45.03       |
|A*     |19       |58838                   |2.35        |
|weighted A*,$\alpha$ = 1.5|19|38679       |1.48        |
|weighted A*,$\alpha$ = 3|21|31118         |1.06        |

#### As a result, we can see that algorithm A* is the fastest. And then BFS and then IDS.                                                                          When alpha increases, the number of states seen and the time decrease, but the answer may not be optimal any more.